server/src/main/java/io/crate/execution/engine/collect/DocValuesGroupByOptimizedIterator.java

```diff
@@ -0,0 +1,428 @@
+import io.crate.expression.symbol.Symbol;
+import io.crate.expression.symbol.Symbols;
+import io.crate.lucene.FieldTypeLookup;
+import io.crate.lucene.LuceneQueryBuilder;
+import io.crate.metadata.DocReferences;
+import io.crate.metadata.Functions;
+import io.crate.metadata.Reference;
+import io.crate.metadata.doc.DocSysColumns;
+import io.crate.metadata.doc.DocTableInfo;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.query.QueryShardContext;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+import static io.crate.execution.dsl.projection.Projections.shardProjections;
+import static io.crate.execution.engine.collect.LuceneShardCollectorProvider.formatSource;
+
+final class DocValuesGroupByOptimizedIterator {
+
+    @Nullable
+    static BatchIterator<Row> tryOptimize(Functions functions,
+                                          IndexShard indexShard,
+                                          DocTableInfo table,
+                                          LuceneQueryBuilder luceneQueryBuilder,
+                                          FieldTypeLookup fieldTypeLookup,
+                                          DocInputFactory docInputFactory,
+                                          RoutedCollectPhase collectPhase,
+                                          CollectTask collectTask) {
+        Collection<? extends Projection> shardProjections = shardProjections(collectPhase.projections());
+        GroupProjection groupProjection = getSinglePartialGroupProjection(shardProjections);
+        if (groupProjection == null) {
+            return null;
+        }
+
+        ArrayList<Reference> docKeyRefs = new ArrayList<>(groupProjection.keys().size());
+        for (var key : groupProjection.keys()) {
+            var keyRef = getKeyRef(collectPhase.toCollect(), key);
+            if (keyRef == null) {
+                return null; // group by on non-reference
+            } else {
+                docKeyRefs.add(keyRef);
+            }
+        }
+
+        ArrayList<Reference> columnKeyRefs = new ArrayList<>(groupProjection.keys().size());
+        for (var keyRef : docKeyRefs) {
```

Comment:
this could be merged-into the loop below and thus would eliminate the `docKeyRefs` list and would avoid another loop.